### PR TITLE
Fix minor issues after introducing incremental JSON_PATCH merge

### PR DESCRIPTION
**Pull request by chrispader**
@marcaaron

### Details

Fixing some issues i encountered while testing the [`JSON_PATCH` merge changes](https://github.com/Expensify/react-native-onyx/pull/238) in the `Expensify/App` repo [in this PR](https://github.com/Expensify/App/pull/23151)

When batching the changes from `mergeQueue`, top-level `null` values may not be omitted. Only when merging the final value, we can remove these `null` values.

Also added back a log line, that was removed in the other PR

### CODE REVIEW

File: lib/Onyx.js
Comment (tgolen) [inline lib/Onyx.js:901]:
```suggestion
    if (!hasChanged) { 
        return Promise.resolve();
     }
```
Can you also please add a comment about _why_ this early return should be here?

File: lib/Onyx.js
Comment (tgolen) [inline lib/Onyx.js:1018]: same

**Review by tgolen — CHANGES_REQUESTED** [review]

File: lib/Onyx.js
Comment (marcaaron) [reply lib/Onyx.js:901]: +1 to the comment

File: lib/Onyx.js
Comment (marcaaron) [inline lib/Onyx.js:852]: Returning `hasChanged` from `broadcastUpdate()` changes it's responsibility a bit. 

Could we maybe pass `hasChanged` as an argument instead? That way the callers already have the value?

File: lib/Onyx.js
Comment (marcaaron) [reply lib/Onyx.js:949]: This is a tricky one. If we are using this method to build the "batch of changes" then don't we actually want to _drop_ those keys from storage and not just set them to "null" or is it the same thing as far as SQLite is concerned?

The only reason to "omit" them from the final object is so they get removed from storage.

Comment (marcaaron) [reply lib/Onyx.js:949]: Maybe it would be clearer to explicitly remove the top level null keys with some brief explanation _after_ we get the "result" (i.e. `modifiedData`) and basically make this method not responsible for modifying the object in that way?

**Review by marcaaron — CHANGES_REQUESTED** [review]

File: lib/Onyx.js
Comment (chrispader) [reply lib/Onyx.js]: Added comments! We returned a resolved promise before too, when the value hasn't changed. That's why i added this back, same for `merge`

**Comment by chrispader** [comment]
@marcaaron @tgolen all review comments adressed! 👍

File: lib/Onyx.js
Comment (tgolen) [inline lib/Onyx.js:903]: This comment should explain _why_ and it currently only explains _what_. Why doesn't it need to be written to storage?

> If the value has not changed, calling Storage.setItem() would be redundant and a waste of performance, so return early instead.

File: lib/Onyx.js
Comment (tgolen) [inline lib/Onyx.js:963]: This comment is no longer valid. It also makes me think that we should still be using `Object.assign()` instead of the spread operator since performance is a very big consideration in this code.

File: lib/Onyx.js
Comment (tgolen) [inline lib/Onyx.js:1015]: Again, please modify the comment to explain _why_ we want to remove all top level keys that are null.

> For objects, the key for null values needs to be removed from the object to ensure the value will get removed from storage completely.

File: lib/Onyx.js
Comment (tgolen) [inline lib/Onyx.js:1025]: Same as above

**Comment by marcaaron** [comment]
+1 to @tgolen's comments about comments :)

>don't we actually want to drop those keys from storage and not just set them to "null" or is it the same thing as far as SQLite is concerned?

Also, I think this is a part that could be better explained in the comments as well. `JSON_PATCH()` behavior will delete a top level key when the value is `null` like in this example...

**Comment by chrispader** [comment]
> Also, I think this is a part that could be better explained in the comments as well. `JSON_PATCH()` behavior will delete a top level key when the value is `null` like in this example...

Ah i understand, didn't know, that that's why we added this manually for web. Gonna further explain this in the comments then!

**Comment by chrispader** [comment]
> If we are using this method to build the "batch of changes" then don't we actually want to drop those keys from storage and not just set them to "null" or is it the same thing as far as SQLite is concerned?

When building the batch of changes, we don't want to omit the null keys, because then this would happen:

Changes like `[{key1: null, key2: null}, {key3: null}]` turn into `modifiedData`: `{}`. Merging the existing object value with `{}` will not update it though, because merging `{}` is the same as merging nothing.

For creating the `modifiedData` we'll want to omit the keys, so they get removed from storage.

File: lib/Onyx.js
Comment (chrispader) [reply lib/Onyx.js:963]: Oh i thought the lodash comment was referring to that we the `fastMerge` instead of `_.merge()`.

According to [this source](https://thecodebarbarian.com/object-assign-vs-object-spread.html), object spreading seems to be the faster choice, especially when we pass an empty object as the target to `Object.assign`, as we do right now.

Comment (tgolen) [reply lib/Onyx.js:963]: I'd be careful referring to that article which is about Node.js. I've done a lot of performance analyzing in the browser and found the spread operator to be much slower than `Object.assign` (which could have just as much to do with our transpiler).

File: lib/Onyx.js
Comment (tgolen) [inline lib/Onyx.js:1017]:
```suggestion
                // On native, SQLite will remove top-level keys that are null. To be consistent, we remove them on web too.
```

**Review by tgolen — CHANGES_REQUESTED** [review]
Getting close!

File: lib/Onyx.js
Comment (chrispader) [reply lib/Onyx.js:963]: Alright! Let's switch back to `Object.assign` then. 👍

Comment (chrispader) [reply lib/Onyx.js:963]: Technically, we don't need the `Object.assign` after all, because fast merge is already copying into a new object:

https://github.com/Expensify/react-native-onyx/blob/d4fc53cc3229808e9c1c542c13ce0089e5893d96/lib/fastMerge.js#L20

Comment (tgolen) [reply lib/Onyx.js:963]: I'm all for removing it entirely!

**Review by tgolen — DISMISSED** [review]

File: lib/Onyx.js
Comment (marcaaron) [inline lib/Onyx.js:951]: The `omitNull` here needs to be removed.

**Review by marcaaron — CHANGES_REQUESTED** [review]
One small thing that needs to be addressed. Looks good - thanks for the changes!

File: lib/Onyx.js
Comment (chrispader) [reply lib/Onyx.js]: right, sorry for that 🥴

File: lib/Onyx.js
Comment (marcaaron) [inline lib/Onyx.js:869]:
```suggestion
```

File: lib/Onyx.js
Comment (marcaaron) [inline lib/Onyx.js:853]:
```suggestion
```

**Review by marcaaron — DISMISSED** [review]

**Comment by marcaaron** [comment]
One last thing 🥲

**Comment by chrispader** [comment]
> One last thing 🥲

fixed!

**Review by mountiny — APPROVED** [review]
Thanks @tgolen and @marcaaron for pushing this forwards.

To unblock the other onyx caching PR, I am going to approve and merge this, given Tim already gave approval and the last changes requested by Marc has been addressed and they were just comments.

**Pull request merged**
